pkg/volume/portworx/portworx.go

```diff
@@ -0,0 +1,386 @@
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.PortworxVolume != nil {
+		return spec.PersistentVolume.Spec.PortworxVolume, spec.ReadOnly, nil
+	}
+
+	return nil, false, fmt.Errorf("Spec does not reference a Portworx Volume type")
+}
+
+// Abstract interface to PD operations.
+type portworxManager interface {
+	// Creates a volume
+	CreateVolume(provisioner *portworxVolumeProvisioner) (volumeID string, volumeSizeGB int, labels map[string]string, err error)
+	// Deletes a volume
+	DeleteVolume(deleter *portworxVolumeDeleter) error
+	// Attach a volume
+	AttachVolume(mounter *portworxVolumeMounter) (string, error)
+	// Detach a volume
+	DetachVolume(unmounter *portworxVolumeUnmounter) error
+	// Mount a volume
+	MountVolume(mounter *portworxVolumeMounter, mountDir string) error
+	// Unmount a volume
+	UnmountVolume(unmounter *portworxVolumeUnmounter, mountDir string) error
+}
+
+// portworxVolume volumes are portworx block devices
+// that are attached to the kubelet's host machine and exposed to the pod.
+type portworxVolume struct {
+	volName string
+	podUID  types.UID
+	// Unique id of the PD, used to find the disk resource in the provider.
+	volumeID string
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager portworxManager
+	// Mounter interface that provides system calls to mount the global path to the pod local path.
+	mounter mount.Interface
+	plugin  *portworxVolumePlugin
+	volume.MetricsProvider
+}
+
+type portworxVolumeMounter struct {
+	*portworxVolume
+	// Filesystem type, optional.
+	fsType string
+	// Specifies whether the disk will be attached as read-only.
+	readOnly bool
+	// diskMounter provides the interface that is used to mount the actual block device.
+	diskMounter *mount.SafeFormatAndMount
+}
+
+var _ volume.Mounter = &portworxVolumeMounter{}
+
+func (b *portworxVolumeMounter) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly: b.readOnly,
+		Managed:  !b.readOnly,
+		// true ?
+		SupportsSELinux: true,
+	}
+}
+
+// Checks prior to mount operations to verify that the required components (binaries, etc.)
+// to mount the volume are available on the underlying node.
+// If not, it returns an error
+func (b *portworxVolumeMounter) CanMount() error {
+	return nil
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (b *portworxVolumeMounter) SetUp(fsGroup *int64) error {
+	return b.SetUpAt(b.GetPath(), fsGroup)
+}
+
+// SetUpAt attaches the disk and bind mounts to the volume path.
+func (b *portworxVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
+	notMnt, err := b.mounter.IsLikelyNotMountPoint(dir)
+	glog.V(4).Infof("Portworx Volume set up: %s %v %v", dir, !notMnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("Cannot validate mountpoint: %s", dir)
+		return err
+	}
+	if !notMnt {
+		return nil
+	}
+
+	if _, err := b.manager.AttachVolume(b); err != nil {
```

Comment:
The benefit of implementing the attacher interface is that the attach/detach operation are handled by the attach/detach controller instead of kubelet. Two benefits of this are 1) credentials required for attach/detach do not need to be available on node machines (for kubelet), and 2) if a node/kubelet becomes inaccessible, the master attach/detach controller can still ensure that volumes are detached (volumes are not stranded). For these reasons, you may want to consider refactoring in the future to implement the attacher interface. `WaitForAttach()` can always be a no-op if it is irrelevant.